client/hotrod-client/src/main/java/org/infinispan/client/hotrod/counter/operation/CompareAndSwapOperation.java

```diff
@@ -28,25 +29,28 @@
    private final long update;
    private final CounterConfiguration counterConfiguration;
 
-   public CompareAndSwapOperation(Codec codec, TransportFactory transportFactory, AtomicInteger topologyId,
-         Configuration cfg, String counterName, long expect, long update, CounterConfiguration counterConfiguration) {
-      super(codec, transportFactory, topologyId, cfg, counterName);
+   public CompareAndSwapOperation(Codec codec, ChannelFactory channelFactory, AtomicInteger topologyId,
+                                  Configuration cfg, String counterName, long expect, long update, CounterConfiguration counterConfiguration) {
+      super(codec, channelFactory, topologyId, cfg, counterName);
       this.expect = expect;
       this.update = update;
       this.counterConfiguration = counterConfiguration;
    }
 
    @Override
-   protected Long executeOperation(Transport transport) {
-      HeaderParams header = writeHeaderAndCounterName(transport, COUNTER_CAS_REQUEST);
-      transport.writeLong(expect);
-      transport.writeLong(update);
-      transport.flush();
+   protected void executeOperation(Channel channel) {
+      ByteBuf buf = getHeaderAndCounterNameBufferAndRead(channel, COUNTER_CAS_REQUEST, 16);
+      buf.writeLong(expect);
+      buf.writeLong(update);
+      channel.writeAndFlush(buf);
```

Comment:
AFAIK, all calls to channel.write need to be done after checking if the channel is writable. If the server is overloaded, the channel might not be writable in which case the client needs to queue up the operations. By overriding `channelWritabilityChanged`, you get a notification when the channel is writable again and you can send the operations.

We had to deal with this when implementing remote events on the server. We had to add back-pressure to the server so that it would not overload the clients. I think the same would be required here, we don't want the clients overloading the server.

Reply:
Good point, I didn't know that. Since I have to apply backpressure to the application thread (which is running this code with the custom channel pool), is it possible to block on the `writeAndFlush` when the channel is not writable?

Reply:
I guess you're talking about sync calls (e.g. put(k, v)) that require some kind of blocking if they can't write? For async calls (e.g. putAsync) you'd not need to block anyway.

Anyway, I guess it might depend on which the the thread is. You definitely don't want to block a thread belonging to Netty. An application thread blocking might be fine, e.g. for sync calls.

@wburns Thoughts?

Reply:
@galderz Yes, the intention is to block only application threads. How else would you apply backpressure? Queuing the write request does not solve anything.
IMO it's OK to block `putAsync()` when the network is saturated. You're not waiting until the server responds, just until you write stuff down the line.
AFAIK netty threads should not call cache.put() unless this is invoked from a listener (we probably discourage that).

Reply:
> IMO it's OK to block putAsync() when the network is saturated. You're not waiting until the server responds, just until you write stuff down the line.

To me blocking putAsync() does not make sense. Even if the network is saturated, it should return early. When the network is no longer saturated, the request can be send, the response received and complete the completable future that putAsync returns.

Reply:
@galderz Would you throw an exception in this case, then?

Reply:
Not sure but don't think so. If anything, the client that waits for the CompletableFuture could get a timeout exception.

Reply:
@galderz To sum our IRC discussion:
* `writeAndFlush` does not block the writer if channel is not writable, it queues the write
* I am strongly convinced that another layer of queuing wouldn't be beneficial
* we should probably throw exception if the channel queues too much

However, as the current code does not allow requests to be multiplexed over channel, there's always at most one in-flight operation on given channel. Therefore I'll defer any checks for overused channel to https://issues.jboss.org/browse/ISPN-8620 where this could really happen. The situation as in server (sending too many events which the client is not handling) does not happen at this moment.